research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+  """The DualNetwork class for the complete model with graph and weights.
+
+  This class can restore the model from saved files, and provide inference for
+  given examples.
+  """
+
+  def __init__(self, save_file, params):
+    """Initialize the dual network from saved model/checkpoints.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+      params: Hyper parameters for the dual network
+    """
+    self.save_file = save_file
+    self.hparams = params
+    self.inference_input = None
+    self.inference_output = None
+    config = tf.ConfigProto()
+    config.gpu_options.allow_growth = True
+    self.sess = tf.Session(graph=tf.Graph(), config=config)
+    self.initialize_graph()
+
+  def initialize_graph(self):
+    with self.sess.graph.as_default():
+      input_features, labels = get_inference_input()
+      estimator_spec = dualnet_model.model_fn(
+          input_features, labels, tf.estimator.ModeKeys.PREDICT, self.hparams)
+      self.inference_input = input_features
+      self.inference_output = estimator_spec.predictions
+      if self.save_file is not None:
+        self.initialize_weights(self.save_file)
+      else:
+        self.sess.run(tf.global_variables_initializer())
+
+  def initialize_weights(self, save_file):
+    """Initialize the weights from the given save_file.
+
+    Assumes that the graph has been constructed, and the save_file contains
+    weights that match the graph. Used to set the weights to a different version
+    of the player without redefining the entire graph.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+    """
+
+    tf.train.Saver().restore(self.sess, save_file)
+
+  def run(self, position, use_random_symmetry=True):
+    """Compute the policy and value output for a given position."""
+    probs, values = self.run_many(
+        [position], use_random_symmetry=use_random_symmetry)
+    return probs[0], values[0]
+
+  def run_many(self, positions, use_random_symmetry=True):
+    processed = list(map(features.extract_features, positions))
+    if use_random_symmetry:
+      syms_used, processed = symmetries.randomize_symmetries_feat(processed)
+    outputs = self.sess.run(
+        self.inference_output, feed_dict={self.inference_input: processed})
+    probabilities, value = outputs['policy_output'], outputs['value_output']
+    if use_random_symmetry:
+      probabilities = symmetries.invert_symmetries_pi(syms_used, probabilities)
+    return probabilities, value
+
+
+def get_inference_input():
+  """Set up placeholders for input features/labels.
+
+  Returns:
+    The features and output tensors that get passed into model_fn.
+  """
+  input_features = tf.placeholder(
+      tf.float32, [None, go.N, go.N, features.NEW_FEATURES_PLANES],
+      name='pos_tensor')
+
+  labels = {
+      'pi_tensor': tf.placeholder(tf.float32, [None, go.N * go.N + 1]),
+      'value_tensor': tf.placeholder(tf.float32, [None])
+  }
```

Comment:
Please comment-- what do each of these represent, what will happen when they get fed into the session?